### DISCUSSION TITLE
Can I enable the --debug after the app was built?

### DISCUSSION THREAD

**Discussion by Thoufak** [pyinstaller/pyinstaller#7256] [Help] [answered]
I'm dealing with a bug which only appears in 5-10% of the builds of my app. Moreover, if I encounter it and try to build my app from the same sources using the same PyInstaller command (actually, changing zero things from my side), the bug will likely disappear. This prevents me from creating a build both with the bug being reproduced and built with --debug option.

So is there any way to turn on the debug logs after the app was built? I guess the debug build uses a different bootloader (run_d.exe?) and perhaps there's a way to replace the bootloader in the ready exe?

**Comment by rokm** [comment, marked as answer]
No, you cannot switch dynamically between debug and non-debug, because they effectively change parts of the built application bundle:

* `--debug bootloader` controls whether regular bootloader executable (`run.exe`) or the one with debug traces (`run_d.exe`) is used. With some effort, you could extract the embedded PKG archive from the non-debug executable, and append it to the `run_d.exe` to obtain executable with bootloader's debug messages enabled (simple append works for Windows; on other OSes, particularly on macOS, it's more complicated). You could also just take the generated PKG from `build/<name>/<name>.pkg` instead of extracting it from the executable. I sometimes use this technique to swap/update the bootloader (e.g., custom version with extra messages) while debugging binary-only builds provided by users in issue reports; not sure how much help would it actually be with your particular problem, though...
* `--debug imports` adds a flag to the generated PKG, which is read by the bootloader to determine whether it should turn on [`Py_VerboseFlag`](https://docs.python.org/3/c-api/init.html#c.Py_VerboseFlag) ([to 1](https://github.com/pyinstaller/pyinstaller/blob/f47d59d0ca4da25386a1d80912f875d908471003/bootloader/src/pyi_pythonlib.c#L210-L212)) before starting embedded python interpreter. This is what turns on all those messages about modules being imported. Toggling this flag requires a rebuild of PKG, so essentially a complete rebuild.
* `--debug noarchive` causes PyInstaller to collect python modules as individual .pyc files instead of putting them into PYZ archive (which is then stored in the PKG archive). Changing this would also require a complete rebuild. This might also subtly change the behavior of the application, because PYZ-embeded python modules are handled by PyInstaller's FrozenImporter, while individual .pyc files are handled by python's built-in importers/loaders.
* `--debug all` turns on all of the above, so by extension, it cannot be toggled without a full rebuild.

**Reply by rokm** [reply]
That said, you could try enabling reproducible builds, by setting `PYTHONHASHSEED` and `SOURCE_DATE_EPOCH`, as described [here](https://pyinstaller.org/en/stable/advanced-topics.html#creating-a-reproducible-build). This *should* result in identical executable being built on each run, assuming that everything else in your environment remains unchanged (so if the problem is indeed with some builds as you describe, you will either always end up with a working build or a defunct one).

**Reply by Thoufak** [reply]
Thank you for a detailed response!
